### PR TITLE
Modified babbage cache control headers 

### PR DESCRIPTION
**Pull request by willcAND**
### What

Modified the proxy cache control response header to prefix the max-age with "public" or "private"

### How to review

Visually inspect code changes, confirm logic matches ticket (DIS-479)

### Who can review

A member of the ONS.

### CODE REVIEW

File: response/writer.go [hunk 1]
```diff
@@ -53,9 +53,15 @@ func writeUnmodifiedResponse(ctx context.Context, w http.ResponseWriter, babbage
...
 func writeResponseWithMaxAge(ctx context.Context, w http.ResponseWriter, babbageResponse *http.Response, maxAge int) {
-	cacheControlValue := fmt.Sprintf("max-age=%d", maxAge)
-	additionalHeaders := map[string]string{
-		"Cache-Control": cacheControlValue,
+	// Prepare the additional headers
+	additionalHeaders := make(map[string]string)
+
+	// Get the original Cache-Control value and modify it to include max-age
+	originalCacheControl := babbageResponse.Header.Get("Cache-Control")
+	if originalCacheControl != "" {
```
Comment (melsner11Ons) [inline response/writer.go:60]: The card mentions you should check the incoming page header for an exact match for "private" or "public"? This above implementation does not do an exact check but it does add max-age in any case?

File: features/cache_control_modification.feature [hunk 2]
```diff
@@ -0,0 +1,18 @@
+Feature: Proxy Modification of Cache-Control Header Based on Release Time
```
Comment (jmgq) [inline features/cache_control_modification.feature]: I think this new feature needs to be refactored, because it's a bit misleading. We're not modifying `Cache-Control` based on the release time. All we need to test is that when Babbage returns a certain `Cache-Control` value ("", "private" or "public"), then we're adding the max-age without overwriting the existing value of `Cache-Control`. Refactoring this will also make the whole feature more readable.

File: response/writer.go [hunk 3]
```diff
@@ -33,7 +33,7 @@ func writeResponse(ctx context.Context, w http.ResponseWriter, babbageResponse *
 
 	// Add any other additional headers
 	for name, value := range additionalHeaders {
-		w.Header().Add(name, value)
+		w.Header().Set(name, value)
```
Comment (jmgq) [inline response/writer.go:36]: How come this change is needed? By doing this, we're no longer adding additional headers, we're potentially overwriting existing ones.

File: response/writer.go (hunk 1 above)
Comment (jmgq) [reply response/writer.go:60]: By the time we reach this code, we know that we need to add `max-age`, it does not matter what is the value of `Cache-Control`, we need to add `max-age`, no matter what. But you are right, we need to add tests for those scenarios. It should be easy to do with the changes I suggested for the feature file.

File: response/writer.go [hunk 4]
```diff
@@ -53,9 +53,15 @@ func writeUnmodifiedResponse(ctx context.Context, w http.ResponseWriter, babbage
 }
 
 func writeResponseWithMaxAge(ctx context.Context, w http.ResponseWriter, babbageResponse *http.Response, maxAge int) {
-	cacheControlValue := fmt.Sprintf("max-age=%d", maxAge)
-	additionalHeaders := map[string]string{
-		"Cache-Control": cacheControlValue,
+	// Prepare the additional headers
+	additionalHeaders := make(map[string]string)
```
Comment (jmgq) [inline response/writer.go]: With the changes you're making to the code, perhaps it means that we need to rename this. It does not make sense to call this `additionalHeaders` anymore, because you're actually overwriting some of them. So I would change the name of that variable (and parameter of `writeResponse`).

Comment (jmgq) [reply response/writer.go]: So something like this:
```go
var cacheControlValue string

originalCacheControl := babbageResponse.Header.Get("Cache-Control")

if originalCacheControl != "" {
	cacheControlValue = fmt.Sprintf("%s, max-age=%d", originalCacheControl, maxAge)
} else {
	cacheControlValue = fmt.Sprintf("max-age=%d", maxAge)
}

headersToUpdate := map[string]string{
	"Cache-Control": cacheControlValue,
}

writeResponse(ctx, w, babbageResponse, headersToUpdate)
```

And then, of course, change the name of the `writeResponse` parameter.

File: response/writer.go (hunk 1 above)
Comment (willcAND) [reply response/writer.go:60]: Yes this is so the logic remains the same when `shouldCalculateMaxAge` returns true with an empty string (https://github.com/ONSdigital/dp-legacy-cache-proxy/blob/develop/response/writer.go#L72) As the ticket is only relating to "public" or "private".

File: features/cache_control_modification.feature [hunk 5]
```diff
@@ -0,0 +1,18 @@
+Feature: Proxy Modification of Cache-Control Header to include max-age when Babbage sets Cache-Control to "public" or "private"
```
Comment (jmgq) [inline features/cache_control_modification.feature]:
```suggestion
Feature: Modification of Cache-Control header to append max-age when Babbage has already set Cache-Control
```

File: features/cache_control_modification.feature [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
```
Comment (jmgq) [inline features/cache_control_modification.feature:6]: You can add a description here, see how it has been done in the other feature files. You could explain what this does in plain English. For instance:

> There are certain instances in which the Proxy needs to set the max-age directive, but Babbage has already set the Cache-Control header (for instance, to "public" or "private"). In these scenarios, the Proxy has to set the max-age directive without overwriting the existing headers.

File: features/cache_control_modification.feature [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  Scenario Outline: Proxy adjusts Cache-Control header based on page release time context
```
Comment (jmgq) [inline features/cache_control_modification.feature]: This isn't accurate. It adjusts the Cache-Control header when the Cache-Control header set by Babbage has certain values.

File: features/cache_control_modification.feature [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    Given Babbage will send the following response:
+      """
+      Mock response from Babbage
+      """
+    And Babbage will set the "Cache-Control" header to "<Babbage-Cache-Control>"
+    And the "/some-path" page <Release-Time-Context>
```
Comment (jmgq) [inline features/cache_control_modification.feature]: This makes this test more complicated than it needs to be. You should remove this parameter, as it does not really affect this test.

File: response/writer.go [hunk 9]
```diff
@@ -53,12 +53,18 @@ func writeUnmodifiedResponse(ctx context.Context, w http.ResponseWriter, babbage
 }
 
 func writeResponseWithMaxAge(ctx context.Context, w http.ResponseWriter, babbageResponse *http.Response, maxAge int) {
-	cacheControlValue := fmt.Sprintf("max-age=%d", maxAge)
-	additionalHeaders := map[string]string{
-		"Cache-Control": cacheControlValue,
+	// Prepare the additional headers
```
Comment (jmgq) [inline response/writer.go]: You can get rid of this comment, as it does not add any value (in fact it is wrong, "additional headers" is the old name).

File: response/writer.go [hunk 10]
```diff
@@ -23,17 +23,17 @@ func WriteResponse(ctx context.Context, w http.ResponseWriter, babbageResponse *
 	}
 }
 
-func writeResponse(ctx context.Context, w http.ResponseWriter, babbageResponse *http.Response, additionalHeaders map[string]string) {
+func writeResponse(ctx context.Context, w http.ResponseWriter, babbageResponse *http.Response, newHeaders map[string]string) {
```
Comment (jmgq) [inline response/writer.go]: `newHeaders` is not accurate enough, but at the same time, I don't know what else to suggest. They're not new headers, because they could already exist in the response. These are headers that we will definitely set in the response, overwriting existing ones. Can you think of a good name for this? If not, we should probably write some Godoc for this function, and explain it there.

Comment (willcAND) [reply response/writer.go]: I've gone with `overrideHeaders` without any documentation explaining the purpose, happy to add some though if you feel its needed

File: features/cache_control_modification.feature (hunk 8 above)
Comment (willcAND) [reply features/cache_control_modification.feature]: removed

File: features/cache_control_modification.feature (hunk 7 above)
Comment (willcAND) [reply features/cache_control_modification.feature]: Updated 👍

File: response/writer.go (hunk 10 above)
Comment (jmgq) [reply response/writer.go]: `overrideHeaders` is a good, descriptive name 👍

File: features/cache_control_modification.feature [hunk 11]
```diff
@@ -0,0 +1,22 @@
+Feature: Modification of Cache-Control header to append max-age when Babbage has already set Cache-Control
+
+    There are certain instances in which the Proxy needs to set the max-age directive, but Babbage has already 
+    set the Cache-Control header (for instance, to "public" or "private"). In these scenarios, 
+    the Proxy has to set the max-age directive without overwriting the existing headers.
+
+  Background:
```
Comment (jmgq) [inline features/cache_control_modification.feature]: Since there is only one scenario in this file, we don't need this. You can simply move this to the scenario itself, and if you reorder the sentences it will read nicely:

Given Babbage will send the following response
And Babbage will set the "Cache-Control" header to "<Babbage-Cache-Control>"
When the Proxy receives a GET request for "/some-path"
Then the response header "Cache-Control" should be "<Babbage-Cache-Control>, max-age=900"

You can also get rid of `And the "/some-path" page was released recently`, as that's not really important for this test. That way, it will return the default cache time, and the test is easier to read, because there is less noise, so it's a lot easier to understand what are we trying to test.

**Review by gedge — APPROVED** [review]
